SQLite gets confused about date types in certain CASE constructions

As [discovered](https://github.com/opensafely-core/ehrql/actions/runs/7416510191/job/20181603781) by Hypothesis, in the following example SQLite returns a string rather than a date:
```python
# example.py

import datetime

from ehrql.query_model.nodes import (
    AggregateByPatient,
    Case,
    SelectTable,
    TableSchema,
    Value,
)
from tests.generative.test_query_model import data_setup


e0 = SelectTable("e0", TableSchema())

population = AggregateByPatient.Exists(e0)

variable = Case(
    {
        Value(False): None,
    },
    default=Value(datetime.date(2000, 1, 1)),
)

data = [
    {"type": data_setup.E0, "patient_id": 1},
]
```

```console
$ just gentest-example-run example.py

AssertionError: Mismatch between in_memory and sqlite
assert equals failed
  [                                    [                                   
    {                                    {                                 
      'patient_id': 1,                     'patient_id': 1,                
      'v': datetime.date(2000, 1, 1),      'v': '2000-01-01',              
    },                                   },                                
  ]                                    ]
```

This is specifically related to having `None` as the only non-default CASE value. Changing the example to the below gives the right type:
```py
    {
        Value(False): Value(datetime.date(1900, 1, 1)),
    },
```

Presumably this is related to the fact that SQLite doesn't actually have a date type but some combination of SQLite/the Python driver/SQLAlchemy conspire to pretend that it does and that in this particular case the conspiracy breaks down.

--- a/ehrql/query_engines/base_sql.py
+++ b/ehrql/query_engines/base_sql.py
@@ -551,7 +551,21 @@ class BaseSQLQueryEngine(BaseQueryEngine):
             default = self.get_expr(node.default)
         else:
             default = None
-        return sqlalchemy.case(*cases, else_=default)
+        expr = sqlalchemy.case(*cases, else_=default)
+        # Temporary workaround for bug in SQLAlchemy. This can be removed when this PR
+        # is merged and released:
+        # https://github.com/sqlalchemy/sqlalchemy/pull/10847
+        expr.type = next(
+            # Find the type of the first non-null case value
+            (
+                value.type
+                for condition, value in expr.whens
+                if not isinstance(value.type, sqlalchemy.types.NullType)
+            ),
+            # Use the type of the `else` clause if there are no matches above
+            expr.else_.type if expr.else_ is not None else sqlalchemy.types.NullType(),
+        )
+        return expr
 
     @get_sql.register(AggregateByPatient.Sum)
     def get_sql_sum(self, node):

